Delete cancelled media from the store

aforcier opened the pull request:
Part of the fix for https://github.com/wordpress-mobile/WordPress-Android/issues/5942.

~We had a `MediaModel.UploadState.CANCELED` state, but it was unused. I opted to drop this entirely and just delete media that are cancelled from the store. If we ever want to support retrying cancelled media, we can re-introduce this; for now, I think this is more straightforward.~

Current WPAndroid cancels media when exiting the editor, but we want to keep those around so we can retry the upload. To accommodate this, I opted to support a `delete` flag (`true` by default) in `CANCEL_MEDIA_UPLOAD`. When set to `false`, the media item will still be cancelled but it won't be deleted from the store - instead it will be set to a `FAILED` state.

This PR also kills the unused `MediaModel.UploadState.CANCELED`, and corrects some behavior in media deletion and in ignoring upload errors for already-deleted media.

cc @mzorz

aforcier commented:
Updated to support both deleting and retaining the cancelled media item, depending on the `delete` flag passed to `CANCEL_MEDIA_UPLOAD`.

WPAndroid-side PR: https://github.com/wordpress-mobile/WordPress-Android/issues/5942 (this PR has breaking changes, so the WPAndroid-side PR should be reviewed (but not merged) before this one is merged).

Ready for review @mzorz!

mzorz commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java:
```diff
@@ -80,7 +80,9 @@ public MediaXMLRPCClient(Dispatcher dispatcher, RequestQueue requestQueue, OkHtt
     @Override
     public void onProgress(MediaModel media, float progress) {
-        notifyMediaProgress(media, Math.min(0.99f, progress), null);
+        if (mCurrentUploadCalls.containsKey(media.getId())) {
```
nice catch here

mzorz commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java:
```diff
@@ -219,10 +221,14 @@ public void onResponse(Call call, Response response) throws IOException {
                 }
+
+                MediaError error = MediaError.fromIOException(e);
+                notifyMediaUploaded(media, error);
```
should we add the same line as [in here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/457/files#diff-4b66763ebf0af33585da417243b8d3a1R341)
`error.message = e.getLocalizedMessage();`

aforcier replied:
Hmm actually that one seems wrong - `fromIOException()` [already calls `e.getLocalizedMessage()`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/60bc5ecbc915998d58fa480f3bd8af457bad824f/fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java#L139). I think I'll make them consistent and remove that from `MediaRestClient` instead 😛

aforcier replied:
Done in 4782097.

mzorz replied:
good call! 🙇

mzorz commented on fluxc/src/main/java/org/wordpress/android/fluxc/persistence/MediaSqlUtils.java:
```diff
@@ -236,11 +236,24 @@ public static MediaModel insertMediaForResult(MediaModel media) {
-                .equals(MediaModelTable.LOCAL_SITE_ID, media.getLocalSiteId())
-                .equals(MediaModelTable.MEDIA_ID, media.getMediaId())
-                .endGroup().endWhere().execute();
+        if (media.getMediaId() == 0) {
```
I'm not sure what zero means here and why it has a different treatment. Can we add a comment explaining this?

aforcier replied:
Good point - I added some comments in 1a5c80e (and also in `insertOrUpdateMedia()` above which has the same switch based on `remote media ID == 0`).

mzorz replied:
tyty

aforcier commented:
Updated in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/457/commits/471feab538aed7d6a48868be9a6de8a629e3f32d to fix a race condition that was causing Mario "Lucky" Zorz's tests to fail.

mzorz commented:
Thanks for all this work @aforcier !

So, 3 "big" things we need to work on to finish this part, from what we’ve been seeing so far:

1) issues with actual cancel sometimes not working correctly 100% of then times.

2) now that this PR adds the possibility to cancel an upload with/without deleting the media, adding specific tests for those 2 cases in a differentiated way (and probably also adding a test that tests FluxC handling of mixed cancelations at the same time).

3) Consistency between MediaRestClient and MediaXMLRPCClient (maybe not part of this PR but found while evaluating it)



### Cancel not working 100% of the times on XMLRPC

I made some modifications to the multiupload/multicancel tests in this branch <[issue/delete-canceled-media-tests](https://github.com/wordpress-mobile/WordPress-FluxC-Android/tree/issue/delete-canceled-media-tests)> as we found in our debugging sessions together they weren’t super reliable across systems/networks and also depended too much on the image being used to upload and the wait time chosen. 
Basically, now the test only starts issuing `cancel` actions when it’s sure there are enough multiple uploads happenings (i.e. won’t issue a cancelation for an upload that we’re not sure to be already in progress). With this, it’s keeping the right balance to make the test work without too much of a problem. If you like that, please feel free to cherry pick commit 1c30e21 and include it in your branch (or do differently if you find there's a better approach to it).

Once that was sorted out, with this I also could confirm another of your findings @aforcier, which I disregarded at first, but evidently has a direct implication for this PR: every now and then, issuing a `CANCEL` action on a progressing upload ends up with a `Socket closed` error, and the corresponding media item is not being deleted from the MediaStore, causing the test assertions to fail.

More information on this is below (please note this only happened for me on XMLRPC, and not all the times):

OK so, every time the test fails (it is not deleting the media item), this is what we see in the logs: 
`media upload failed: java.net.SocketException: Socket closed`

```
06-10 07:15:16.292 25478-25491/org.wordpress.android.fluxc.example D/WordPress-API: Dispatching action: MediaAction-CANCEL_MEDIA_UPLOAD
06-10 07:15:16.308 25478-25566/org.wordpress.android.fluxc.example D/dalvikvm: GC_FOR_ALLOC freed 401K, 6% free 9217K/9784K, paused 6ms, total 6ms
06-10 07:15:16.324 25478-25562/org.wordpress.android.fluxc.example D/dalvikvm: GC_FOR_ALLOC freed 390K, 6% free 9225K/9784K, paused 6ms, total 6ms
06-10 07:15:16.328 25478-25555/org.wordpress.android.fluxc.example D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
06-10 07:15:16.328 25478-25562/org.wordpress.android.fluxc.example D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
06-10 07:15:16.344 25478-25566/org.wordpress.android.fluxc.example D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
06-10 07:15:16.360 25478-25553/org.wordpress.android.fluxc.example D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
06-10 07:15:16.364 25478-25563/org.wordpress.android.fluxc.example D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
06-10 07:15:16.380 25478-25542/org.wordpress.android.fluxc.example D/WordPress-MEDIA: Canceled in-progress upload: /test1.jpg
06-10 07:15:16.380 25478-25542/org.wordpress.android.fluxc.example D/WordPress-MEDIA: mediaXMLRPCClient: removed id: 1 from current uploads, remaining: 4
06-10 07:15:16.384 25478-25542/org.wordpress.android.fluxc.example D/WordPress-API: Dispatching action: MediaAction-CANCELED_MEDIA_UPLOAD
06-10 07:15:16.428 25478-25583/org.wordpress.android.fluxc.example E/WordPress-MEDIA: cancelled media, total image count: 6
06-10 07:15:16.432 25478-25482/org.wordpress.android.fluxc.example D/dalvikvm: GC_CONCURRENT freed 459K, 6% free 9269K/9784K, paused 35ms+1ms, total 88ms
06-10 07:15:16.432 25478-25553/org.wordpress.android.fluxc.example D/dalvikvm: WAIT_FOR_CONCURRENT_GC blocked 58ms
06-10 07:15:16.432 25478-25553/org.wordpress.android.fluxc.example W/WordPress-MEDIA: media upload failed: java.net.SocketException: Socket closed
[…]
```

OTOH every time the upload and cancel works, we see this:

`media upload failed: java.io.IOException: Canceled`

```
06-10 07:19:07.204 26319-26332/org.wordpress.android.fluxc.example D/WordPress-API: Dispatching action: MediaAction-CANCEL_MEDIA_UPLOAD
06-10 07:19:07.216 26319-26323/org.wordpress.android.fluxc.example D/dalvikvm: GC_CONCURRENT freed 465K, 6% free 9216K/9796K, paused 8ms+7ms, total 49ms
06-10 07:19:07.216 26319-26390/org.wordpress.android.fluxc.example D/dalvikvm: WAIT_FOR_CONCURRENT_GC blocked 18ms
06-10 07:19:07.224 26319-26372/org.wordpress.android.fluxc.example D/WordPress-MEDIA: Canceled in-progress upload: /test1.jpg
06-10 07:19:07.224 26319-26372/org.wordpress.android.fluxc.example D/WordPress-MEDIA: mediaXMLRPCClient: removed id: 1 from current uploads, remaining: 4
06-10 07:19:07.224 26319-26372/org.wordpress.android.fluxc.example D/WordPress-API: Dispatching action: MediaAction-CANCELED_MEDIA_UPLOAD
06-10 07:19:07.224 26319-26381/org.wordpress.android.fluxc.example W/WordPress-MEDIA: media upload failed: java.io.IOException: Canceled
06-10 07:19:07.224 26319-26370/org.wordpress.android.fluxc.example E/WordPress-MEDIA: cancelled media, total image count: 4
```

Let’s investigate this a bit further to make tests work 100% of time - let me know if you can’t reproduce and I’ll get on it.


### Adding tests for delete=`true/false`  flag in `CancelMediaUpload`

Given this PR introduces a new treatment / differentiation in logic, we need to provide a new test that takes into account _not deleting the media_. I suggest (but haven’t thoroughly tested) pretty much copying the `testUploadMultipleImagesAndCancel()` test and make these changes:


1st change: the cancellation should be done with the `false` flag:
               ` CancelMediaPayload payload = new CancelMediaPayload(sSite, media, false);`

(we’ll have to refactor `uploadMultipleMedia`  a bit)

2nd change: in `testUploadMultipleImagesAndCancel()`, change this line:
```
//        assertEquals(mUploadedIds.size(), mMediaStore.getSiteImageCount(sSite));
```
for this block:
```
        // Only completed uploads should exist in the store
        List<MediaModel> siteImages = mMediaStore.getSiteImages(sSite);
        int remoteImagesCount = 0;
        if (siteImages != null) {
             for (MediaModel image : siteImages) {
                 if (image.getUploadState().equals(MediaModel.UploadState.UPLOADED.toString())) {
                     remoteImagesCount++;
                 }
             }
        }
        assertEquals(mUploadedIds.size(), remoteImagesCount);
```


### Consistency between MediaREstClient and MediaXMLRPCCLient

I know it’s not directly part of this PR but we should try to make both clients have the same logic / code, and in that I found this:
`removeCallFromCurrentUploadsMap()` method  is available in MediaXMLRPCClient but not in MediaRestClient, for example. I’m sure there are a couple more things to check, with the help of a vertical split screen comparison.


That's pretty much it for now (well, we know we've put a lot in here!), hopefully when this is sorted out we'll have a much stronger library! 
As usual, I ❤️ your work and working with you!
🙇

aforcier commented:
> 1. Issues with actual cancel sometimes not working correctly 100% of the time

As discussed, we'll attack this one in a separate PR after we figure out just what is causing the `java.net.SocketException: Socket closed` for you.

> 2. Adding tests for delete=true/false flag in CancelMediaUpload

Added in 3650b47 (note that single-upload cancel with `delete = false` was already added in [testCancelImageUpload()](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/457/files#diff-d1032009dc29eb8d9f7b6de3425c3172R217). I also added some extra verification of upload state to the other multi-upload tests.

> 3. Consistency between MediaRestClient and MediaXMLRPCClient

I started doing this but the changes I'd like to make aren't trivial and I'm not convinced this should be done in this PR as it's already quite large. I suggest cleaning that up in a separate PR.

mzorz commented:
Thank you @aforcier ! Agreed on points 1 and 2 to be tackled on PRs of their own each. Will follow up on nbr 2 here

mzorz commented:
> 2. Adding tests for delete=true/false flag in CancelMediaUpload
This works OK, now testing https://github.com/wordpress-mobile/WordPress-Android/issues/5942. Will update there accordingly 👍
